hazelcast/src/main/java/com/hazelcast/ringbuffer/impl/RingbufferService.java

```diff
@@ -101,29 +128,113 @@ public void shutdown(boolean terminate) {
+        ringbuffer.getStore().instrument(nodeEngine);
+        partitionContainers.put(namespace, ringbuffer);
+        return ringbuffer;
+    }
+
+    public boolean hasContainer(int partitionId, ObjectNamespace namespace) {
+        final ConcurrentMap<ObjectNamespace, RingbufferContainer> partitionContainers = containers.get(partitionId);
+        return partitionContainers != null && partitionContainers.get(namespace) != null;
+    }
+
+    private Map<ObjectNamespace, RingbufferContainer> getOrCreateRingbufferContainers(int partitionId) {
+        Map<ObjectNamespace, RingbufferContainer> partitionContainer = containers.get(partitionId);
+        if (partitionContainer == null) {
+            containers.putIfAbsent(partitionId, new ConcurrentHashMap<ObjectNamespace, RingbufferContainer>());
+        }
+        return containers.get(partitionId);
+    }
+
+    public RingbufferConfig getRingbufferConfig(String name) {
+        Config config = nodeEngine.getConfig();
+        return config.getRingbufferConfig(getConfigName(name));
+    }
+
+    public static ObjectNamespace getRingbufferNamespace(String name) {
+        return new DistributedObjectNamespace(SERVICE_NAME, name);
+    }
+
+    public int getRingbufferPartitionId(String ringbufferName) {
+        final Data partitionAwareData =
+                nodeEngine.getSerializationService().toData(ringbufferName, StringPartitioningStrategy.INSTANCE);
+        return nodeEngine.getPartitionService().getPartitionId(partitionAwareData);
+    }
+
+    public void addRingbuffer(int partitionId, RingbufferContainer ringbuffer, RingbufferConfig config) {
+        checkNotNull(ringbuffer, "ringbuffer can't be null");
+        final SerializationService serializationService = nodeEngine.getSerializationService();
+        ringbuffer.init(config, serializationService, nodeEngine.getConfigClassLoader());
+        ringbuffer.getStore().instrument(nodeEngine);
+        getOrCreateRingbufferContainers(partitionId).put(ringbuffer.getNamespace(), ringbuffer);
+    }
+
     @Override
     public void beforeMigration(PartitionMigrationEvent partitionMigrationEvent) {
     }
 
+    /**
+     * {@inheritDoc}
+     * <p>
+     * Here we will replicate only ringbuffer containers for the ringbuffer service. Other services are responsible
+     * for migrating their ringbuffer containers on migration events. The main reason is because of fine-grained
+     * anti entropy (replication). The partition data might not be migrated in full but rather in part.
+     * For instance, the partition data for a single map might not be in sync between replicas but the ringbuffer
+     * data might be in sync. This means that the map data will be replicated but the ringbuffer data will not be replicated
+     * which can cause differences between ringbuffer data for some services.
+     * In another words, each service is responsible for replicating all of it's data, including data in other services.
```

Comment:
I don't think this is exactly true. A namespace can be shared between services and while migrating a namespace, migration system asks all participating services to replicate data for that namespace. But still each service is responsible for replicating it's own managed data.